### PR TITLE
file upload max file accept closure

### PR DESCRIPTION
**Pull request by m-wardany**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes/no
| Tests pass?   | yes
| Fixed issues  |  #14031

### CODE REVIEW

File: framework/validators/FileValidator.php [hunk 1]
```diff
@@ -444,11 +450,12 @@ public function getClientOptions($model, $attribute)
...
-            $options['maxFiles'] = $this->maxFiles;
-            $options['tooMany'] = $this->formatMessage($this->tooMany, [
+            $maxFiles = $this->maxFiles instanceof \Closure? call_user_func($this->maxFiles, $model, $attribute): $this->maxFiles ;
+            $options['maxFiles'] = $maxFiles;
+            $options['tooMany'] = Yii::$app->getI18n()->format($this->tooMany, [
                 'attribute' => $label,
-                'limit' => $this->maxFiles,
-            ]);
+                'limit' => $maxFiles,
+            ], Yii::$app->language);
```
Comment (MysteryDragon) [inline framework/validators/FileValidator.php]: It's not related to this issue, isn't it?

File: framework/validators/FileValidator.php [hunk 2]
```diff
@@ -185,7 +186,12 @@ public function init()
      */
     public function validateAttribute($model, $attribute)
     {
-        if ($this->maxFiles != 1) {
+        $maxFiles = $this->maxFiles instanceof \Closure? call_user_func($this->maxFiles, $model, $attribute): $this->maxFiles ;
+        
+        if($maxFiles == 0 && $this->maxFiles instanceof \Closure){
```
Comment (MysteryDragon) [inline framework/validators/FileValidator.php]: PSR-2 mistakes.
$maxFiles < 0 => OK?

File: framework/validators/FileValidator.php [hunk 3]
```diff
@@ -185,7 +186,12 @@ public function init()
...
+            $this->addError($model, $attribute, $this->tooMany, ['limit' => $maxFiles]);
+        }
+        elseif ($maxFiles != 1) {
```
Comment (MysteryDragon) [inline framework/validators/FileValidator.php]: PSR-2 mistake

File: framework/validators/FileValidator.php [hunk 4]
```diff
@@ -68,10 +68,11 @@ class FileValidator extends Validator
      */
     public $maxSize;
     /**
-     * @var int the maximum file count the given attribute can hold.
+     * @var int|Closure the maximum file count the given attribute can hold.
      * Defaults to 1, meaning single file upload. By defining a higher number,
      * multiple uploads become possible. Setting it to `0` means there is no limit on
      * the number of files that can be uploaded simultaneously.
+     * you may define it as closure to calculate the required number of attached files.
```
Comment (MysteryDragon) [inline framework/validators/FileValidator.php]: First letter must be in upper case.

File: framework/validators/FileValidator.php [hunk 5]
```diff
@@ -185,7 +186,12 @@ public function init()
      */
     public function validateAttribute($model, $attribute)
     {
-        if ($this->maxFiles != 1) {
+        $maxFiles = $this->maxFiles instanceof \Closure? call_user_func($this->maxFiles, $model, $attribute): $this->maxFiles ;
```
Comment (MysteryDragon) [inline framework/validators/FileValidator.php]: I think `$result = $condition ? $success : $fail;` formatting will be more appropriate in Yii2. Partially it's also related with PSR-2.

**Comment by MysteryDragon** [comment]
I guess unit test will be required.

**Comment by samdark** [comment]
Closed for the reason described in the issue.

**Pull request closed without merging**